XiEditor/LineCache.swift

```diff
@@ -172,55 +177,160 @@ class LineCache {
         nInvalidBefore = newInvalidBefore
         lines = newLines
         nInvalidAfter = newInvalidAfter
+
         if height < oldHeight {
             inval.addRange(start: height, end: oldHeight)
         }
         return inval
     }
 
-    /// Return ranges of invalid lines within the given range
-    func computeMissing(_ first: Int, _ last: Int) -> [(Int, Int)] {
-        var result: [(Int, Int)] = []
-        let last = min(last, height)  // lines past the end aren't considered missing
-        guard first < last else {
-            Swift.print("compute missing called with first > last")
-            return result
-        }
-        
-        for ix in first..<last {
-            // could optimize a bit here, but unlikely to be important
-            if ix < nInvalidBefore || ix >= nInvalidBefore + lines.count || lines[ix - nInvalidBefore] == nil {
-                if result.count == 0 || result[result.count - 1].1 != ix {
-                    result.append((ix, ix + 1))
-                } else {
-                    result[result.count - 1].1 = ix + 1
-                }
+    /// The set of lines which contain cursors.
+    var cursorInval: InvalSet {
+        let inval = InvalSet()
+        for (i, line) in lines.enumerated() {
+            if line?.containsCursor ?? false {
+                inval.addRange(start: i + nInvalidBefore, n: 1)
             }
         }
-        return result
+        return inval
+    }
+}
+
+//TODO: use os_unfair_lock_t on 10.12+
+
+/// A safe wrapper around a system lock.
+class Mutex {
+    fileprivate var mutex = pthread_mutex_t()
+
+    init() {
+        pthread_mutex_init(&mutex, nil)
+    }
+
+    deinit {
+        pthread_mutex_destroy(&mutex)
+    }
+
+    func lock() {
+        pthread_mutex_lock(&mutex)
+    }
+
+    func unlock() {
+        pthread_mutex_unlock(&mutex)
+    }
+
+    func isLocked() -> Bool {
+        if pthread_mutex_trylock(&mutex) == 0 {
+            defer { pthread_mutex_unlock(&mutex) }
+            return false
+        } else {
+            return true
+        }
+    }
+}
+
+/**
+ A cache of lines representing a document in xi-core. The cache is updated based
+ on deltas from the core.
+ 
+ - Note: To facilitate smooth scrolling, updates to the LineCache are expected
+ to arrive on a dedicated thread. When drawing, lines are fetched through the
+ `blockingGet(lines:)` method, which will block for some maximum ammount of time
+ waiting for the lines to arrive from xi-core.
+ */
+class LineCache {
+
+    /// A semaphore we use to wake up the main thread if it is blocking missing lines
+    fileprivate let waitingForLines = DispatchSemaphore(value: 0)
+    /// A dispatch queue used to synchronize access to the underlying state
+    fileprivate let queue = DispatchQueue(label: "io.xi-editor.cache-sync-queue")
+    /// A marker used to indicate if the main thread is blocked.
+    fileprivate let blockFlag = Mutex()
```

Comment:
My intention is to communicate that this isn't being used as a mutex in the traditional sense (it isn't guarding a shared resource; the queue does that) but just as a way of communicating to the update thread that it needs to signal. 

(I'll continue and address the subsequent comment as well)

My initial implementation of this was using `NSCondition`, but I struggled to make that work, specifically around timeouts. Apple is strongly pushing `libdispatch` based approaches to concurrency, and the best documentation & examples I could find were based on dispatch queues and dispatch semaphores; but although I thought it was doable, I couldn't figure out a way to make that approach work without introducing some additional primitive. 

I agree that there is a possible race here; however the worst case in the race scenario (where the main thread doesn't have the lock when the update thread checks, but grabs it before the update returns) is that the main thread blocks for the full timeout interval or until the next update arrives, whichever is less.

Another approach I could consider is trying to implement this using `pthread_cond_t`, but then I have to implement my own timeout logic, which will be tricky.

_Ideally_ there's some solution here that just involves the semaphores and I just couldn't find it.

Reply:
If it's _strictly_ being used for signalling, it should be an atomic boolean instead of a mutex. But I'm not sure those are first class in Swift.

Another thought I had, what about putting the semaphore inside the mutex? My thinking is that this would guarantee each wait is balanced by exactly one signal. But I'm spitballing; it feels like there probably is a good concurrency primitive for this and we just don't know what it is yet.

Reply:
I've rewritten this to take the approach you've suggested of always having the wait call happen while holding the lock, which solves the problem with the race.